### PR TITLE
Collate Support for the SQL

### PR DESCRIPTION
**Pull request by adkhare**
We have been facing issues with select statements not able to handle `collate` operators. 

This PR is to add the feature for `collate` operator support

### CODE REVIEW

File: go/vt/sqlparser/ast.go [hunk 1]
```diff
@@ -1801,7 +1802,7 @@ func (node ColIdent) CompliantName() string {
 	return compliantName(node.val)
 }
 
-// Lowered returns a lower-cased column name.
+// Lowered returns a lowezr-cased column name.
```
Comment (harshit-gangal) [inline go/vt/sqlparser/ast.go]: z added by mistake in "lowezr". can be removed.

**Comment by adkhare** [comment]
@harshit-gangal addressed the above

**Comment by adkhare** [comment]
@sougou Please can you review and provide feedback. Would like to have this merged by monday as there is a pressing need internally from the product.

File: go/vt/sqlparser/sql.y [hunk 2]
```diff
@@ -769,6 +770,10 @@ compare:
   {
     $$ = NullSafeEqualStr
   }
+| COLLATE
```
Comment (sougou) [inline go/vt/sqlparser/sql.y]: This shouldn't be required. Can you remove it and see if things work as intended?

File: go/vt/tabletserver/endtoend/compatibility_test.go [hunk 3]
```diff
@@ -18,7 +18,7 @@ import (
 var point12 = "\x00\x00\x00\x00\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf0?\x00\x00\x00\x00\x00\x00\x00@"
 
 func TestCharaterSet(t *testing.T) {
-	qr, err := framework.NewClient().Execute("select * from vitess_test where intval=1", nil)
+	qr, err := framework.NewClient().Execute("select * from vitess_test where intval=1 order by charval collate utf8_general_ci asc", nil)
```
Comment (sougou) [inline go/vt/tabletserver/endtoend/compatibility_test.go]: Let's revert this test change. I'm now convinced that `collate` doesn't affect character escaping. There's no need to change this.

File: go/vt/sqlparser/sql.y [hunk 4]
```diff
@@ -869,6 +874,10 @@ value_expression:
   {
     $$ = &BinaryExpr{Left: $1, Operator: JSONUnquoteExtractOp, Right: $3}
   }
+| value_expression COLLATE value_expression
```
Comment (sougou) [inline go/vt/sqlparser/sql.y]: Add a test for this to parse_test.go (TestValid).

**Comment by adkhare** [comment]
@sougou added the test for parse_test. Also, made changes as per your recommendations

File: go/vt/sqlparser/parse_test.go [hunk 5]
```diff
@@ -610,6 +610,8 @@ func TestValid(t *testing.T) {
 		input: "select /* GE true */ 1 from t where a >= true",
 	}, {
 		input: "select /* GE false */ 1 from t where a >= false",
+	}, {
+		input: "select * from t order by a collate utf8_general_ci asc",
```
Comment (bbeaudreault) [inline go/vt/sqlparser/parse_test.go]: it's up to @sougou but I feel like it's worth adding a test case for each of the forms the COLLATE keyword can be used: http://dev.mysql.com/doc/refman/5.7/en/charset-collate.html

**Comment by adkhare** [comment]
Just added some more changes to the tests as the parser converts all the SQL's to lowercase

**Pull request merged**